### DISCUSSION TITLE
What does "never_convert_lossy_files" mean exactly?

### DISCUSSION THREAD

**Discussion by 2V3EvG4LMJFdRe** [beetbox/beets#5275] [General]
This option of the convert plugin has made my head scratch before - [and others too](https://github.com/beetbox/beets/issues/3124). It goes as follows:

`never_convert_lossy_files: Cross-conversions between lossy codecs—such as mp3, ogg vorbis, etc.—makes little sense as they will decrease quality even further. If set to yes, lossy files are always copied. Default: no.`

That first part I agree with. But then the wording gets confusing quickly, files are always "copied" - does that mean "copied" or "converted"? And then, the default "no" for a never_convert_lossy_files wording just sends my head on a spiral.

Furthermore, I'm almost certain the default option transcodes mp3 -> mp3.

Bottom line is, do I fill in "yes" or "no" if I only want it to convert lossless -> lossy?

**Comment by RollingStar** [comment]
Some general tips which I use for most open source projects:

1. Make sure you're on the latest/beta branch (our "stable" is quite old)
2. search the latest docs for your issue (here, "never_convert_lossy_files" is a nice keyword)
3. search the code on github (you can see the magic likely happens here https://github.com/beetbox/beets/blob/1b594790ca3c156c807cc09b85945373c198c272/beetsplug/convert.py#L98 )

In this case, there's also a (failed) pull request you can reference. https://github.com/beetbox/beets/pull/3361/files

All that said:

```
def should_transcode(item, fmt):
    """Determine whether the item should be transcoded as part of
    conversion (i.e., its bitrate is high or it has the wrong format).
    """
    no_convert_queries = config["convert"]["no_convert"].as_str_seq()
    if no_convert_queries:
        for query_string in no_convert_queries:
            query, _ = parse_query_string(query_string, Item)
            if query.match(item):
                return False
    if config["convert"]["never_convert_lossy_files"] and not (
        item.format.lower() in LOSSLESS_FORMATS
    ):
        return False
```

If you've set 

config["convert"]["never_convert_lossy_files"] to True,

and the item.format.lower() is not in LOSSLESS_FORMATS,

should_transcode returns False. https://github.com/beetbox/beets/blob/1b594790ca3c156c807cc09b85945373c198c272/beetsplug/convert.py#L88

> do I fill in "yes" or "no" if I only want it to convert lossless -> lossy?

You can see it's rather byzantine but I think the answer is yes/True.

**Reply by 2V3EvG4LMJFdRe** [reply]
Thank you for the throughout explanation, I'm fairly sure it has confused others before. The default mode transcodes say mp3 > mp3, weird! Another thing I would recommend that I did was compare with Audacity the waveform of a pre and post-processed .mp3 file to make sure it had stayed the same.

**Reply by RollingStar** [reply]
I used to be steadfast on lossy-> lossy, but I've been doing it for years in portable listening and never noticed a problem. (Usually mp3->opus @ 96k). Beets, being a library manager, should have some way to have a canonical copy and a 'converted' lossy copy. as long as you have the source (even if it's lossy), lossy-lossy shouldn't be a big deal. I believe all this only applies to `convert` plugin but I think you can do an in-place `convert` and erase the original files.

**Reply by 2V3EvG4LMJFdRe** [reply]
Hmm that would explain it. I have just started looking into the opus format, it's a very interesting development that had passed me by. Pretty amazing to see low kbps like that offer comparable results to 320 mp3.
